### PR TITLE
Method not allowed

### PR DESCRIPTION
**Pull request by LeSpocky**
### Identify the Bug

#240 

### Description of the Change

To be able to send along an additional header like `Allow: GET, POST` with an HTTP 405 (Method Not Allowed) response, the `httpserver::http_resource` class had to be extended to expose the allowed methods already stored in that object. Those can be used in the webserver then to build the necessary header line and attach it to the response.

### Alternate Designs

The proposed version was selected based on discussion in the bug report #240.

### Possible Drawbacks

- public API change (`httpserver::http_resource` had to be changed)
- Additional calculation overhead

### Verification Process

Tests were added:

- unit tests for the extension of the class `httpserver::http_resource`
- integration tests for webserver to verify header is actually send in case

### Release Notes

- Class `http_resource` now allows to query the currently set allowed methods
- Webserver now sends 'Allow:' header along with 405 (Method Not Allowed) responses, to comply with HTTP standard (RFC 7231)

### CODE REVIEW

File: test/integ/basic.cpp [hunk 1]
```diff
@@ -1021,6 +1021,33 @@ LT_BEGIN_AUTO_TEST(basic_suite, url_with_regex_like_pieces)
...
+    curl_easy_setopt(curl, CURLOPT_URL, "localhost:8080/base");
+    curl_easy_setopt(curl, CURLOPT_HTTPGET, 1L);
+    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writefunc);
+    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &s);
+    curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, headerfunc);
+    curl_easy_setopt(curl, CURLOPT_HEADERDATA, &ss);
+    res = curl_easy_perform(curl);
+    LT_ASSERT_EQ(res, 0);
+    long http_code = 0;
+    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &http_code);
```
Comment (LeSpocky) [inline test/integ/basic.cpp:1044]: cpplint complains about this, however according to https://curl.se/libcurl/c/CURLINFO_RESPONSE_CODE.html *long* is the correct type for `curl_easy_getinfo()` here.

**Comment by LeSpocky** [comment]
The other thing cpplint complains about is header ordering, which will be easy to fix.

File: test/integ/basic.cpp (hunk 1 above)
Comment (LeSpocky) [reply test/integ/basic.cpp:1044]: Used `int64_t` like elsewhere in that file now. 🤷🏼‍♂️

**Comment by LeSpocky** [comment]
> The other thing cpplint complains about is header ordering, which will be easy to fix.

Done that and force pushed a new version. Lets see what CI says now?!

File: src/webserver.cpp [hunk 2]
```diff
@@ -601,6 +601,15 @@ MHD_Result webserver::finalize_answer(MHD_Connection* connection, struct details
...
             } else {
                 mr->dhrs = method_not_allowed_page(mr);
+
+                vector<string> allowed_methods = hrm->get_allowed_methods();
+                if (allowed_methods.size() > 0) {
+                    string header_value = allowed_methods[0];
+                    for (auto it = allowed_methods.cbegin() + 1; it != allowed_methods.cend(); ++it) {
+                        header_value += ", " + (*it);
+                    }
+                    mr->dhrs->with_header("Allow", header_value);
```
Comment (etr) [inline src/webserver.cpp:611]: Can we use: https://github.com/etr/libhttpserver/blob/263f2bdef4d5f8887a46c19fa08695170ad679f7/src/http_utils.cpp#L135

**Review by etr — COMMENTED** [review]
It all looks good, only one comment to use the constant defined in http_utils for the "Allow" header.

File: src/webserver.cpp (hunk 2 above)
Comment (LeSpocky) [reply src/webserver.cpp]: Sure, no problem. Force pushed the updated branch.

**Pull request merged**